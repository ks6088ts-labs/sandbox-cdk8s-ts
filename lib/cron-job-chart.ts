import { Construct } from 'constructs'
import { Chart, ChartProps } from 'cdk8s'
import { KubeCronJob } from '../imports/k8s'
import { RestartPolicy } from 'cdk8s-plus-23'

export class CronJobChart extends Chart {
  constructor(scope: Construct, id: string, props: ChartProps = {}) {
    super(scope, id, props)

    new KubeCronJob(this, 'cronjob', {
      metadata: {
        name: 'periodic-task',
      },
      spec: {
        // crontab guru: https://crontab.guru/
        schedule: '*/1 * * * *',
        successfulJobsHistoryLimit: 10,
        jobTemplate: {
          spec: {
            completions: 3,
            parallelism: 1,
            backoffLimit: 0,
            template: {
              spec: {
                containers: [
                  {
                    name: 'hello',
                    image: 'busybox',
                    command: [
                      '/bin/sh',
                      '-c',
                      'date; echo Hello from the Kubernetes cluster',
                    ],
                  },
                ],
                restartPolicy: RestartPolicy.ON_FAILURE,
              },
            },
          },
        },
      },
    })
  }
}
