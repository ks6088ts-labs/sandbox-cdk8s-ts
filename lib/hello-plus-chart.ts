import { Construct } from "constructs";
import * as kplus from "cdk8s-plus-23";
import { Chart, ChartProps } from "cdk8s";

export class HelloPlusChart extends Chart {
  constructor(scope: Construct, id: string, props: ChartProps = {}) {
    super(scope, id, props);

    new kplus.Deployment(this, "Deployment", {
      replicas: 3,
      containers: [
        {
          image: "ubuntu",
        },
      ],
    });
  }
}
