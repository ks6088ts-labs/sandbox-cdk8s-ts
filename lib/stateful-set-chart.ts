import { Construct } from 'constructs'
import { Chart, ChartProps } from 'cdk8s'

// imported constructs
import {
  KubeDeployment,
  KubeService,
  IntOrString,
  KubeStatefulSet,
  Quantity,
} from '../imports/k8s'
import * as kplus from 'cdk8s-plus-23'

export class StatefulSetChart extends Chart {
  constructor(scope: Construct, id: string, props: ChartProps = {}) {
    super(scope, id, props)

    const label = { app: 'stateful-set' }
    const quantity = Quantity.fromString('1G')

    new KubeStatefulSet(this, 'stateful-set', {
      metadata: {
        labels: label,
      },
      spec: {
        serviceName: 'stateful-set',
        replicas: 3,
        selector: {
          matchLabels: label,
        },
        template: {
          metadata: {
            labels: label,
          },
          spec: {
            containers: [
              {
                name: 'nginx',
                image: 'nginx:mainline-alpine-slim',
                volumeMounts: [
                  {
                    name: 'www',
                    mountPath: '/usr/share/nginx/html',
                  },
                ],
              },
            ],
          },
        },
        volumeClaimTemplates: [
          {
            metadata: {
              name: 'www',
            },
            spec: {
              accessModes: [kplus.PersistentVolumeAccessMode.READ_WRITE_ONCE],
              resources: {
                requests: { storage: quantity },
              },
            },
          },
        ],
      },
    })

    new KubeService(this, 'service', {
      spec: {
        type: 'LoadBalancer',
        ports: [{ port: 80, targetPort: IntOrString.fromNumber(80) }],
        selector: label,
      },
    })

    new KubeDeployment(this, 'deployment', {
      spec: {
        replicas: 2,
        selector: {
          matchLabels: label,
        },
        template: {
          metadata: { labels: label },
          spec: {
            containers: [
              {
                name: 'site',
                image: 'nginx:mainline-alpine-slim',
                ports: [{ containerPort: 80 }],
              },
            ],
          },
        },
      },
    })
  }
}
