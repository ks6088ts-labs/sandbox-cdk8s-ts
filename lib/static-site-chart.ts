import { Construct } from 'constructs'
import { Chart, ChartProps } from 'cdk8s'

// imported constructs
import { KubeDeployment, KubeService, IntOrString } from '../imports/k8s'

export class StaticSiteChart extends Chart {
  constructor(scope: Construct, id: string, props: ChartProps = {}) {
    super(scope, id, props)

    const label = { app: 'static-site' }

    new KubeService(this, 'service', {
      spec: {
        type: 'LoadBalancer',
        ports: [{ port: 80, targetPort: IntOrString.fromNumber(80) }],
        selector: label,
      },
    })

    new KubeDeployment(this, 'deployment', {
      spec: {
        replicas: 2,
        selector: {
          matchLabels: label,
        },
        template: {
          metadata: { labels: label },
          spec: {
            containers: [
              {
                name: 'site',
                image: 'nginx:mainline-alpine-slim',
                ports: [{ containerPort: 80 }],
              },
            ],
          },
        },
      },
    })
  }
}
