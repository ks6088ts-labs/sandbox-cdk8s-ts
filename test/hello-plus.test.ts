import { HelloPlusChart } from "../lib/hello-plus-chart";
import { Testing } from "cdk8s";

describe("Placeholder", () => {
  test("Empty", () => {
    const app = Testing.app();
    const chart = new HelloPlusChart(app, "test-chart");
    const results = Testing.synth(chart);
    expect(results).toMatchSnapshot();
  });
});
