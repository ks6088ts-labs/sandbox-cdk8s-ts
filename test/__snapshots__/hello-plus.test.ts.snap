// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Placeholder Empty 1`] = `
Array [
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "test-chart-deployment-c8cfd183",
    },
    "spec": Object {
      "minReadySeconds": 0,
      "progressDeadlineSeconds": 600,
      "replicas": 3,
      "selector": Object {
        "matchLabels": Object {
          "cdk8s.io/metadata.addr": "test-chart-Deployment-c8550d6d",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": "25%",
          "maxUnavailable": "25%",
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "cdk8s.io/metadata.addr": "test-chart-Deployment-c8550d6d",
          },
        },
        "spec": Object {
          "automountServiceAccountToken": false,
          "containers": Array [
            Object {
              "image": "ubuntu",
              "imagePullPolicy": "Always",
              "name": "main",
              "resources": Object {
                "limits": Object {
                  "cpu": "1500m",
                  "memory": "2048Mi",
                },
                "requests": Object {
                  "cpu": "1000m",
                  "memory": "512Mi",
                },
              },
              "securityContext": Object {
                "allowPrivilegeEscalation": false,
                "privileged": false,
                "readOnlyRootFilesystem": true,
                "runAsNonRoot": true,
              },
            },
          ],
          "dnsPolicy": "ClusterFirst",
          "restartPolicy": "Always",
          "securityContext": Object {
            "fsGroupChangePolicy": "Always",
            "runAsNonRoot": true,
          },
          "setHostnameAsFQDN": false,
        },
      },
    },
  },
]
`;
